### PR TITLE
Use explicit type when not apparent or intrinsic

### PR DESCRIPTION
**Pull request by jcouv**
Setting `csharp_style_var_elsewhere = false:warning` preference for compiler (so that it doesn't interfere with work-in-progress) and applying fix all.

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Awesome.  out of curiosity, how long did it take for fix-all to do its work?

**Comment by CyrusNajmabadi** [comment]
Also, i spot checked this, and didn't notice anything innapropriate.  Did you discover any examples where the IDE wanted to convert to an explicit type, but you though the type was apparent?

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler We can discuss at tomorrow's team meeting. 
An easier and more incremental alternative would be to use setting `info` and not apply fix all.

File: src/Compilers/.editorconfig [hunk 1]
```diff
@@ -3,4 +3,4 @@
 [*.cs]
 csharp_style_var_for_built_in_types = false:none
```
Comment (sharwell) [inline src/Compilers/.editorconfig]: :memo: This one should also be `false:warning`

**Comment by sharwell** [comment]
> An easier and more incremental alternative would be to use setting info and not apply fix all.

IDE analyzers have no impact on build. We can set it to warning even without fixing anything.

File: src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs [hunk 2]
```diff
@@ -216,12 +216,12 @@ private static bool IsReducedSelectOrGroupByClause(SelectOrGroupClauseSyntax sel
...
             }
 
-            var selectorIdentifier = ((IdentifierNameSyntax)selectOrGroupExpression).Identifier;
+            SyntaxToken selectorIdentifier = ((IdentifierNameSyntax)selectOrGroupExpression).Identifier;
 
             SyntaxToken sourceIdentifier;
             QueryBodySyntax containingBody;
 
-            var containingQueryOrContinuation = selectOrGroupClause.Parent.Parent;
+            CSharpSyntaxNode containingQueryOrContinuation = selectOrGroupClause.Parent.Parent;
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: ❓ How did this happen?

**Comment by jcouv** [comment]
@CyrusNajmabadi I'm not sure how long it took (I left it running and did something else). Maybe 10-15 minutes for getting the preview and another 10-15 for applying the change? (VS hung after that, but the files were saved)

**Comment by jcouv** [comment]
@sharwell I think the plan is to enforce editorconfig preferences during build soon. I think we also treat warnings as errors in CI. So "warning" setting would effectively become "error" at that point, I think.

File: src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: Hum, actually it builds fine from VS and with `msbuild /v:m /m Compilers.sln`. But I get the same error as CI when I run `build.cmd`... That's mysterious

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: Turns out this file is linked into `CSharpFeatures.csproj` too, and that's where it failed (in VS and first command-line, I was only building `Compilers.sln`).

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: > Turns out this file is linked into CSharpFeatures.csproj too, and that's where it failed (in VS and first command-line, I was only building Compilers.sln).

Interesting.  Question:

Did you only have Compilers.sln open in VS when you did the fix?  Or did you have Roslyn.sln open?  If the latter, then we should open a bug on this.

Thanks!

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: I did it with only Compilers.sln open.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: Ah ok.

**Comment by jcouv** [comment]
From discussion with @jaredpar, we're not quite ready to add warning or error enforcement in editor config, because we lack an enforcement mechanism in CI to prevent new violations. The result would be that someone making an edit via notepad or with different local style preferences in VS could check in some code that would break the rest of the team, which is very painful.

I will close this PR for now, and ramp up the editor config to "info" enforcement for now instead. Once we have the enforcement mechanism in CI, we'll re-open the discussion.
Thanks

**Comment by CyrusNajmabadi** [comment]
Everything you said made sense.  But it's unclear why you would close this PR.  Couldn't you just change to ":info" and then just still merge this PR in?

It seems like that would still be good as it would both address the problems now, and make it clearer in the code what teh style is for anyone who makes future changes.  Right now, there are so many usages of 'var' in inappropriate locations, that anyone coming in to edit think that's ok.  If we remove them, and only keep the allowed forms, that will make the style clearer, even if ti's not explicitly enforced.

**Comment by jcouv** [comment]
@CyrusNajmabadi I think it will be easier to justify the churn and merge costs once we have an end-to-end solution. Apparently there was some broader discussion on the transition effort which landed on that approach.

**Comment by sharwell** [comment]
> The result would be that someone making an edit via notepad or with different local style preferences in VS could check in some code that would break the rest of the team, which is very painful.

VS does not enforce warning or error either. It is a UI layer feature only.

**Comment by sharwell** [comment]
@gafter I submitted #25634 as an alternative

**Pull request closed without merging**